File: core/src/main/scala/tripaint/effects/MotionBlurEffect.scala
package tripaint.effects

import tripaint.ColorLookup
import tripaint.color.Color
import tripaint.coords.GlobalPixCoords

class MotionBlurEffect(radius: Int) extends LocalEffect {
  def name: String = "Motion blur"

  private val radiusSq = radius.toDouble * radius

  override protected def predicate(
      image: ColorLookup,
      here: GlobalPixCoords
  )(coords: GlobalPixCoords, color: Color): Boolean = {
    here.y == coords.y && math.pow(here.x - coords.x, 2) <= radiusSq * 1.5
  }

  override protected def weightedColor(image: ColorLookup, here: GlobalPixCoords)(
      coords: GlobalPixCoords
  ): (Double, Color) = {
    (math.exp(-2 * math.pow(here.x - coords.x, 2) / radiusSq), image.lookup(coords).get)
  }
}

File: core/src/main/kotlin/tripaint/effects/MotionBlurEffect.kt
package tripaint.effects

import tripaint.ColorLookup
import tripaint.color.Color
import tripaint.coords.GlobalPixCoords
import kotlin.math.exp
import kotlin.math.pow

class MotionBlurEffect(radius: Int) : LocalEffect() {
    override fun name(): String = "Motion blur"

    private val radiusSq = radius.toDouble() * radius

    override fun predicate(
        image: ColorLookup,
        here: GlobalPixCoords
    ): (coords: GlobalPixCoords, color: Color) -> Boolean {
        return { coords, _ -> here.y == coords.y && (here.x - coords.x).toDouble().pow(2.0) <= radiusSq * 1.5 }
    }

    override fun weightedColor(image: ColorLookup, here: GlobalPixCoords): (
    coords: GlobalPixCoords
    ) -> Pair<Double, Color> {
        return { coords -> Pair(exp(-2 * (here.x - coords.x).toDouble().pow(2.0) / radiusSq), image.lookup(coords)!!) }
    }
}
